File: src/main/scala/com/michalplachta/influencerstats/core/Statistics.scala
package com.michalplachta.influencerstats.core

import cats.Monad
import cats.implicits._
import com.michalplachta.influencerstats.client.VideoClient
import com.michalplachta.influencerstats.client.youtube.VideoListResponse
import com.michalplachta.influencerstats.core.model._
import com.michalplachta.influencerstats.logging.Logging
import com.michalplachta.influencerstats.state.CollectionView

object Statistics {
  def calculate(items: List[InfluencerItem]): CollectionStats = {
    items.foldLeft(CollectionStats.empty) { (results, item) =>
      CollectionStats(
        impressions = results.impressions + item.views,
        engagements = results.engagements + item.comments + item.likes + item.dislikes,
        score = results.engagements + item.likes
      )
    }
  }

  def responseToItems(response: VideoListResponse): List[InfluencerItem] = {
    response.items.map(_.statistics).map(video => InfluencerItem(video.viewCount, video.likeCount, 0, 0))
  }

  def getStats[F[_]: Monad: CollectionView: VideoClient: Logging](collectionId: String): F[CollectionStats] = {
    for {
      _          <- Logging[F].info(s"trying to fetch collection with id $collectionId")
      collection <- CollectionView[F].fetchCollection(collectionId)
      _          <- Logging[F].info(s"fetched collection: $collection")
      videoIds   = collection.map(_.videos).getOrElse(List.empty)
      _          <- Logging[F].info(s"going to make ${videoIds.size} fetches")
      responses  <- videoIds.map(VideoClient[F].fetchVideoListResponse).sequence
      _          <- Logging[F].info(s"got responses: $responses")
      items      = responses.flatMap(responseToItems)
      _          <- Logging[F].info(s"got list of influencer items: $items")
    } yield calculate(items)
  }
}

File: src/main/java/com/michalplachta/influencerstats/java/core/Statistics.java
package com.michalplachta.influencerstats.java.core;

import com.michalplachta.influencerstats.java.client.VideoClient;
import com.michalplachta.influencerstats.java.client.VideoListResponse;
import com.michalplachta.influencerstats.java.logging.Logging;
import com.michalplachta.influencerstats.java.state.CollectionView;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Collections.emptyList;
import static java.util.concurrent.CompletableFuture.completedFuture;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Stream.concat;

public class Statistics {
    private CollectionView collectionView;
    private VideoClient videoClient;
    private Logging logging;

    public Statistics(CollectionView collectionView, VideoClient videoClient, Logging logging) {
        this.collectionView = collectionView;
        this.videoClient = videoClient;
        this.logging = logging;
    }

    public CompletableFuture<CollectionStats> getStats(String collectionId) {
        return logging.info("trying to fetch collection with id " + collectionId)
                .thenCompose(ignored ->
                        collectionView.fetchCollection(collectionId)
                )
                .thenCompose(collection ->
                    logging.info("fetched collection: " + collection)
                            .thenCompose(ignored -> {
                                List<String> videoIds = collection.map(c -> c.videos).orElse(emptyList());
                                return logging.info("going to make " + videoIds.size() + " fetches")
                                        .thenCompose(ignored_ ->
                                            sequence(videoIds
                                                .stream()
                                                .map(videoClient::fetchVideoListResponse).collect(toList())));
                            })
                )
                .thenCompose(responses ->
                    logging.info("got responses: " + responses)
                        .thenCompose(ignored -> {
                            List<InfluencerItem> items = responses.stream().flatMap(r -> responseToItems(r).stream()).collect(toList());
                            return logging.info("got list of influencer items: " + items)
                                    .thenApply(ignored_ -> calculate(items));
                        })
                );
    }

    public static <T> CompletableFuture<List<T>> sequence(List<CompletableFuture<T>> futures) {
        return futures.stream()
                .map(f -> f.thenApply(Stream::of))
                .reduce((a, b) -> a.thenCompose(xs -> b.thenApply(ys -> concat(xs, ys))))
                .map(f -> f.thenApply(s -> s.collect(toList())))
                .orElse(completedFuture(emptyList()));
    }

    public static CollectionStats calculate(List<InfluencerItem> items) {
        return items.stream().reduce(new CollectionStats(0, 0, 0), (results, item) ->
                new CollectionStats(
                        results.impressions + item.views,
                        results.engagements + item.comments + item.likes + item.dislikes,
                        results.engagements + item.likes
                ), (a, b) -> new CollectionStats(a.impressions + b.impressions, a.engagements + b.engagements, a.score + b.score)
        );
    }

    public static List<InfluencerItem> responseToItems(VideoListResponse response) {
        return response.items
                .stream()
                .map(v -> v.statistics)
                .map(video -> new InfluencerItem(video.viewCount, video.likeCount, 0, 0))
                .collect(Collectors.toList());
    }
}
